model/jpa/src/main/java/org/keycloak/models/jpa/session/PersistentClientSessionEntity.java

```diff
@@ -30,15 +30,16 @@
  * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
  */
 @NamedQueries({
-        @NamedQuery(name="deleteClientSessionsByRealm", query="delete from PersistentClientSessionEntity sess where sess.userSessionId IN (select u.userSessionId from PersistentUserSessionEntity u where u.realmId = :realmId)"),
+        @NamedQuery(name="deleteClientSessionsByRealm", query="delete from PersistentClientSessionEntity sess where sess.userSessionId IN (select u.userSessionId from PersistentUserSessionEntity u where u.realmId = :realmId) and sess.offline = coalesce(:offline, sess.offline) "),
         @NamedQuery(name="deleteClientSessionsByClient", query="delete from PersistentClientSessionEntity sess where sess.clientId = :clientId"),
         @NamedQuery(name="deleteClientSessionsByExternalClient", query="delete from PersistentClientSessionEntity sess where sess.clientStorageProvider = :clientStorageProvider and sess.externalClientId = :externalClientId"),
         @NamedQuery(name="deleteClientSessionsByClientStorageProvider", query="delete from PersistentClientSessionEntity sess where sess.clientStorageProvider = :clientStorageProvider"),
         @NamedQuery(name="deleteClientSessionsByUser", query="delete from PersistentClientSessionEntity sess where sess.userSessionId IN (select u.userSessionId from PersistentUserSessionEntity u where u.userId = :userId)"),
         @NamedQuery(name="deleteClientSessionsByUserSession", query="delete from PersistentClientSessionEntity sess where sess.userSessionId = :userSessionId and sess.offline = :offline"),
         @NamedQuery(name="deleteExpiredClientSessions", query="delete from PersistentClientSessionEntity sess where sess.userSessionId IN (select u.userSessionId from PersistentUserSessionEntity u where u.realmId = :realmId AND u.offline = :offline AND u.lastSessionRefresh < :lastSessionRefresh)"),
         @NamedQuery(name="findClientSessionsByUserSession", query="select sess from PersistentClientSessionEntity sess where sess.userSessionId=:userSessionId and sess.offline = :offline"),
-        @NamedQuery(name="findClientSessionsByUserSessions", query="select sess from PersistentClientSessionEntity sess where sess.offline = :offline and sess.userSessionId IN (:userSessionIds) order by sess.userSessionId")
+        @NamedQuery(name="findClientSessionsByUserSessions", query="select sess from PersistentClientSessionEntity sess where sess.offline = :offline and sess.userSessionId IN (:userSessionIds) order by sess.userSessionId"),
+        @NamedQuery(name="findClientSessionsCountByClient", query="select count(sess) from PersistentClientSessionEntity sess where sess.offline = :offline and sess.clientId = :clientId")
```

Comment:
This query should be backed by index.

Reply:
The query `findClientSessionsCountByClient` requires a full table scan for the generated query:
```sql
-- PersistentClientSessionEntity.findClientSessionsCountByClient
select count(persistent0_.CLIENT_ID) as col_0_0_
from OFFLINE_CLIENT_SESSION persistent0_
where persistent0_.OFFLINE_FLAG='1'
  and persistent0_.CLIENT_ID='dce1b319-cac8-41af-bf11-123b768060c1'
 ```
![image](https://user-images.githubusercontent.com/314690/107160038-13036d00-6994-11eb-88e1-b9684349ccf4.png)

With the index:
```sql
create index IDX_OFFLINE_CSS_PRELOAD on offline_client_session (offline_flag, client_id);
```
Created via 
```xml
    <changeSet author="keycloak" id="13.0.0-KEYCLOAK-11019">
        <createIndex tableName="OFFLINE_CLIENT_SESSION" indexName="IDX_OFFLINE_CSS_PRELOAD">
            <column name="OFFLINE_FLAG" type="VARCHAR(4)"/>
            <column name="CLIENT_ID" type="VARCHAR(36)"/>
        </createIndex>
    </changeSet>
```
The index is used and the query is much faster.
![image](https://user-images.githubusercontent.com/314690/107160040-1a2a7b00-6994-11eb-8b3f-58bfbe28f318.png)

Reply:
@thomasdarimont Thanks for the changes and adding the tests .

@hmlnarik Does the index introduced in the commit https://github.com/keycloak/keycloak/pull/7722/commits/02b6e62e1614118957f7a7b5ccea1a5f06c2ee55 work for you? Could it have side-effect on the performance when introducing index?

Reply:
The indices could be improved, see below. The analysis of how to improve the indices is [here](https://docs.google.com/spreadsheets/d/14nk2i7bG0dCeI0MLqANoxEXX1VIehyfOZ5qFQIXoCrs/edit?usp=sharing)